### DISCUSSION TITLE
Requesting cosmetic changes to the code generation output

### DISCUSSION THREAD

**Discussion by lukaseder** [jOOQ/jOOQ#17724] [FAQ]
Code style and formatting is a controversial topic. Every company and even every developer within a company would like to format their code in their own "best" style.

jOOQ's code generator is no different. It produces its own "unique" style that may or may not please the eye of the consumer. It is only natural to wish to customise it to some extent, and even to request feature from jOOQ regarding customisation.

Some customisation features are reasonable, to some extent. E.g.:

- Text block usage: https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-textblocks/
- Whitespace: https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-whitespace/

Other customisations affect functionality more than style, so they're not really part of this discussion:

- `@Generated` Annotation: https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-generate-annotations/
- Setter style: https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-generate-fluent-setters/
- And much more

Our general stance on customisation of code generation style follows these rules:

- If it's a stylistic "bug," then we'll fix it. Example bugs include:
  - There are obvious indentation errors or other types of excess whitespace, such as between Javadoc and method declaration
  - There are inconsistent styles (e.g. `method( )` vs `method()`)
- If it's a stylistic "preference," then we're unlikely going to fix it, unless it's a simple fix, and a strategic one. Example preferences that aren't simple include:
  - Curly braces on the same line or on a different line
  - Some linter warns about some "bad" usage of some token (but it's not an error). This includes all sorts of "warnings" that fail compilation when `-Werror` is used to compile jOOQ code. The solution here is: simply don't do that.

There is gray area, of course. E.g. we occasionally add `@SuppressWarnings` to our generated output, because `javac` doesn't allow for customising its warnings very well. But if warnings arise from purely optional linting, then the solution is simply not to lint.

Look at jOOQ code like you would look at byte code. You don't really care about the style of generated byte code, so why care about generated jOOQ code? There are far more interesting improvements we can make in jOOQ than improved cosmetics in code generation output.

**Comment by agentgt** [comment]
I think using the word "style" is kind of nebulous.

That is there is formatting which is what I think you are largely referring to and then there is "style" which is more like naming and pattern conventions (e.g. getters prefixed with get or withers or suffix vs prefix extending an abstract interface).

And this is important because "formatting" can be handled at build time with things like spotless and a whole plethora of plugins.

Thus for the generated code base one just has to make their build run the formatter after jooq generates the code. 

If the generated code is checked in then even more so why one should let their build time formatter do the work. 

So I just don't think formatting stuff is something you should worry about and this includes the bizarre folks that want curly start braces on their own line (kind of joking but it the thing I hate most about C# code 😄 ... at least Java got that convention right!). Let them run their formatters.

**Reply by lukaseder** [reply]
I said: *"Other customisations affect functionality more than style, so they're not really part of this discussion"*. Getter naming affects adherence to this or that well-known (or lesser-known) convention, such as JavaBeans or Kotlin-whatever. If we don't follow the convention, then interoperability is affected, and that's just a bug.

I believe my examples about "style" are clear with respect to what I don't wish to spend time on? I'll be happy to add more examples as I encounter them in bug reports.

I like your example about the possibility of running a formatter on generated code, which would solve all of these issues thoroughly.

**Reply by lukaseder** [reply]
Note, I just added this FAQ so that I can link to it from any future issue requesting cosmetic changes to generated code.

**Reply by agentgt** [reply]
> I believe my examples about "style" are clear with respect to what I don't wish to spend time on? I'll be happy to add more examples as I encounter them in bug reports.

They mostly are and kind of skim read as I have a headache at the moment.

I was just hoping to maybe help you make that distinction clear by perhaps avoiding the use of the word "style" altogether. 

For example while I gave the naming convention as "style" you are correct that even that is not a good example as naming convention functionally changes things (because of reflection or whatever). Here we have a better phrase for that "naming convention".

Style basically means nothing. You can define it but largely what you are talking about appears to have a better word "format".

The closest thing I can think of "style" like it is perhaps used in writing (e.g. elements of style by strunk) are things like using functional programming instead of imperative programming to reach the same result.

And like I said before the distinction is important because you cannot automate "style" corrections like you can formatting.

Anyway it is not big deal.
